test/widgets/test_notify.py

```diff
@@ -0,0 +1,173 @@
+MESSAGE_2, NOTIFICATION_2 = notification("Urgent Message",
+                                         "This is not a test!",
+                                         urgency=2,
+                                         timeout=10000)
+MESSAGE_3, NOTIFICATION_3 = notification("Low priority",
+                                         "Windows closed unexpectedly",
+                                         urgency=0)
+
+
+@pytest.mark.skipif(
+    shutil.which("notify-send") is None,
+    reason="notify-send not installed."
+    )
+@pytest.mark.usefixtures("dbus")
+def test_notifications(manager_nospawn, minimal_conf_noscreen, monkeypatch):
+    notify.Notify.timeout_add = log_timeout
+    widget = notify.Notify(foreground_urgent=URGENT,
+                           foreground_low=LOW,
+                           default_timeout=DEFAULT_TIMEOUT)
+    config = minimal_conf_noscreen
+    config.screens = [
+        libqtile.config.Screen(
+            top=Bar([widget], 10)
+        )
+    ]
+
+    manager_nospawn.start(config)
+    obj = manager_nospawn.c.widget["notify"]
+
+    # Send first notification and check time and display time
+    notif_1 = [NS]
+    notif_1.extend(NOTIFICATION_1)
+    subprocess.run(notif_1)
+    assert obj.info()["text"] == MESSAGE_1
+
+    _, timeout = obj.eval("self.delay")
+    assert timeout == "5.0"
+
+    # Send second notification and check time and display time
+    notif_2 = [NS]
+    notif_2.extend(NOTIFICATION_2)
+    subprocess.run(notif_2)
+    assert obj.info()["text"] == MESSAGE_2.format(colour=URGENT)
+
+    _, timeout = obj.eval("self.delay")
+    assert timeout == "10.0"
+
+    # Send third notification and check time and display time
+    notif_3 = [NS]
+    notif_3.extend(NOTIFICATION_3)
+    subprocess.run(notif_3)
+    assert obj.info()["text"] == MESSAGE_3.format(colour=LOW)
+
+    _, timeout = obj.eval("self.delay")
+    assert timeout == str(DEFAULT_TIMEOUT)
+
+    # Navigation tests
+
+    # Hitting next while on last message should not change display
+    obj.next()
+    assert obj.info()["text"] == MESSAGE_3.format(colour=LOW)
+
+    # Show previous
+    obj.prev()
+    assert obj.info()["text"] == MESSAGE_2.format(colour=URGENT)
+
+    # Show previous
+    obj.prev()
+    assert obj.info()["text"] == MESSAGE_1
+
+    # Show previous while on first message should stay on first
+    obj.prev()
+    assert obj.info()["text"] == MESSAGE_1
+
+    # Show next
+    obj.next()
+    assert obj.info()["text"] == MESSAGE_2.format(colour=URGENT)
+
+    # Toggle display (clear)
+    obj.toggle()
+    assert obj.info()["text"] == ""
+
+    # Toggle display - restoring display shows last notification
+    obj.toggle()
+    assert obj.info()["text"] == MESSAGE_3.format(colour=LOW)
+
+    # Clear the dispay
+    obj.clear()
+    assert obj.info()["text"] == ""
+
+    # Show the display
+    obj.display()
+    assert obj.info()["text"] == MESSAGE_3.format(colour=LOW)
+
+# untested: 92 - play sound
```

Comment:
What does the 92 here mean?

Reply:
It's to do with the test coverage. It means this line is showing as untested:
https://github.com/qtile/qtile/blob/5ea61eb817791500851b30108e5696631cffc5ec/libqtile/widget/notify.py#L92

I was aiming for 100% coverage where possible.

Reply:
Ah cool. If that line of code moves to a different line if notify.py gets edited, will the change in coverage result be picked up as a test failure?

Reply:
No. It was a note for me as I was writing the test (I started with all the lines that pytest said was untested and removed them as I incorporated them into the test).

Line numbers changing will have no impact on the test.

Reply:
Do we leave it in?

Reply:
Not if you think it's confusing.